### PR TITLE
Documents Tagging architecture decisions

### PR DESCRIPTION
**Pull request by pomegranited**
This PR documents the decisions that arose from https://github.com/openedx/modular-learning/issues/60.

It also builds on https://github.com/openedx/openedx-learning/pull/54 and enforces [ADR 7. Tagging app structure](https://github.com/open-craft/openedx-learning/blob/jill/tagging-data-arch-adr/docs/decisions/0007-tagging-app.rst) by moving the tagging app to a separate package outside of `openedx_learning`, called `openedx_tagging`. The models in this app will be altered and aligned with the other ADRs as part of subsequent PRs.

**Testing instructions**

1. When running the automated tests, check that the `openedx_tagging` test runs.
1. Check the added ADRs for clarity and correctness.

**Supporting information**

Closes https://github.com/openedx/modular-learning/issues/60

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by ChrisChV** [comment]
@pomegranited Great work! :+1: 

I think we need also to specify how will the link with the contents be handled. I have seen that ContenTag is mentioned in the ADRs, but I feel that it has to be specified more. Specify that we have to create subclasses of ObjectTag to link the content, like ContentTag or the class to link the learning core content.

File: docs/decisions/0009-tagging-administrators.rst
Comment (bradenmacdonald) [inline docs/decisions/0009-tagging-administrators.rst:25]: Per OEP-9, these three rules that you listed above can be defined using [rules](https://github.com/dfunckt/django-rules)... with #1 defined in `openedx_tagging` and #2 and #3 defined in Studio. I _think_ that what that gets us is that any application can then use the standard Django API to query things like `user.has_perm('openedx_tagging.edit_taxonomy', taxonomy)` and get the right result without needing to be aware of how those permissions are defined or even where they are defined (in Studio).

File: docs/decisions/0010-taxonomy-enable-context.rst
Comment (bradenmacdonald) [inline docs/decisions/0010-taxonomy-enable-context.rst:17]: I think this needs to be clarified. It sounds like this is saying "the taxonomy can be enabled for all contexts (forced on for all courses, regardless of other settings) when `enabled=True` or disabled for all contexts (forced off for all courses, regardless of other settings) when `enabled=False`". We either need a third state (can be selectively enabled/disabled at other levels when this is set to `None`), or make this into either a "global disable" or "global enable" boolean, but not both. I hope that makes sense.

File: openedx_tagging/core/tagging/models.py
Comment (bradenmacdonald) [inline openedx_tagging/core/tagging/models.py:30]: I think some namespacing spec is required here. 

What if I want to tag a user, which has a numeric ID? Should we specify the format for this field? Perhaps that it's always an "Opaque Key" serialized as a string, or that if it's not an opaque key, it's a similarly prefixed string like `"user:15"` ?

File: docs/decisions/0010-taxonomy-enable-context.rst
Comment (pomegranited) [reply docs/decisions/0010-taxonomy-enable-context.rst]: Yep, that makes sense @bradenmacdonald . Does this look clearer? https://github.com/open-craft/openedx-learning/blob/jill/tagging-data-arch-adr/docs/decisions/0010-taxonomy-enable-context.rst

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:30]: I'd rather add a short string `content_type` field than prefix the `content_id` -- or in terms of the [`ObjectTag` ](https://docs.google.com/document/d/13zfsGDfomSTCp_G-_ncevQHAb4Y4UW0d_6N8R2PdHlA/edit#heading=h.o6fm1hktwp7b) class that will ultimately be the base class for `ContentTag`, I'd rather use `object_type` and `object_id` fields.

File: docs/decisions/0009-tagging-administrators.rst
Comment (pomegranited) [reply docs/decisions/0009-tagging-administrators.rst:25]: That would be really cool @bradenmacdonald , but can django-rules handle chaining multiple rule definitions together?

Ref [rules.RuleSet](https://github.com/dfunckt/django-rules#instance-methods-1):

* `add_rule` raises a KeyError if another rule by that name already exists.
* `set_rule` sets the rule, regardless of whether one already exists

But I don't see how to combine them?

**Comment by pomegranited** [comment]
Thank you for your feedback @ChrisChV  @bradenmacdonald  -- I've addressed what I can from your comments, and from watching the meeting on 6 June. @ormsbee did you have any comments or changes needed here?

I guess I'm also not sure how much of the [data model document](https://docs.google.com/document/d/13zfsGDfomSTCp_G-_ncevQHAb4Y4UW0d_6N8R2PdHlA/edit#heading=h.ogslfhqqejyh) needs to be enshrined in ADRs, and how much can be done when we do the implementation.. Is there anything missing that should be in here?

File: docs/decisions/0010-taxonomy-enable-context.rst
Comment (ormsbee) [inline docs/decisions/0010-taxonomy-enable-context.rst:27]: If the course hasn't enabled all taxonomies, but it has enabled this specific one, shouldn't it be shown for the course?

File: setup.py
Comment (ormsbee) [inline setup.py:74]: Ack, sorry about this, and thank you for fixing it.

**Review by ormsbee — COMMENTED** [review]
Minor question about taxonomy enablement. It doesn't need to block merging.

File: docs/decisions/0009-tagging-administrators.rst
Comment (bradenmacdonald) [reply docs/decisions/0009-tagging-administrators.rst:25]: @pomegranited Isn't [combining rules](https://github.com/dfunckt/django-rules#combining-predicates) what you mean? It supports combining rules with "and", "or", "not", etc.

File: docs/decisions/0010-taxonomy-enable-context.rst
Comment (bradenmacdonald) [reply docs/decisions/0010-taxonomy-enable-context.rst:27]: I think this doesn't allow enabling taxonomies at the course level (yet)... I misunderstood this at first too, but I think `tax.enabled_for(course.org)` means `tax.enabled_for(org)` (where `org` comes from the `course`), not `tax.enabled_for(course | org)` as I first read this.

**Comment by bradenmacdonald** [comment]
Thanks. I like how this is shaping up!

File: docs/decisions/0009-tagging-administrators.rst
Comment (pomegranited) [reply docs/decisions/0009-tagging-administrators.rst:25]: @bradenmacdonald Yes, we can combine rules logically and explicitly -- what I meant was, there's no magic provided by django or django-rules to chain these calls to `set_rule`.

It's totally doable, it just takes a little care. For example: 

1. `openedx_tagging` calls `set_rule('openedx_tagging.edit_taxonomy', can_edit_taxonomy)` where `can_edit_taxonomy` checks e.g `user.is_staff | user.is_superuser`. This is fine, `openedx_tagging` can stand alone.
2. `openedx.features.tagging.SystemTaxonomy` might want to call `set_rule('openedx_tagging.edit_taxonomy', can_edit_taxonomy)` too, where its `can_edit_taxonomy` disallows editing for all SystemTaxonomy no matter who you are.
3. `openedx.features.tagging.OrgTaxonomy` might want to call `set_rule('openedx_tagging.edit_taxonomy', can_edit_taxonomy)` too, where its `can_edit_taxonomy` checks `openedx_tagging.can_edit_taxonomy` and the user's rights against the OrgTaxonomy.org.

But they can't all pile on and call `set_rule`, because the most recently called `set_rule` wins.

So to make sure `user.has_perm('openedx_tagging.edit_taxonomy', taxonomy)` returns the right thing, the `openedx.features.tagging` rules must be aware of its different Taxonomy subclasses, and explicitly combine with the `openedx_tagging` rules where needed.

File: openedx_tagging/core/tagging/models.py
Comment (pomegranited) [reply openedx_tagging/core/tagging/models.py:30]: @bradenmacdonald  Thoughts on ^?

**Comment by pomegranited** [comment]
@bradenmacdonald @ormsbee  Thanks for your comments!  I'm pretty pleased with it too :)

If you're happy with what's here, could one of you approve and squash+merge?
Or if there's more that needs to be in here, please let me know? I'm starting implementation this week.

File: docs/decisions/0009-tagging-administrators.rst
Comment (bradenmacdonald) [reply docs/decisions/0009-tagging-administrators.rst:25]: @pomegranited Ah, I see. Yeah it's too bad there's not a nice API for combining those. maybe we can add a lightweight way to do that?

File: openedx_tagging/core/tagging/models.py
Comment (bradenmacdonald) [reply openedx_tagging/core/tagging/models.py:30]: Yep, that sounds good! As long as it's namespaced in some way.

**Review by bradenmacdonald — APPROVED** [review]

File: docs/decisions/0009-tagging-administrators.rst
Comment (pomegranited) [reply docs/decisions/0009-tagging-administrators.rst:25]: @bradenmacdonald Sure, will see.. I'll start by trying to use consistent convention, and if it ends up being abstractable into something we can reuse, that would be awesome :) Thanks for pointing me to OEP-9 here, and clarifying the end goal of these permission rules.

**Comment by pomegranited** [comment]
Thanks for your 👍 @bradenmacdonald  :)  I don't have merge rights here, so could you or @ormsbee merge if it's ready? Thank you!

**Comment by openedx-webhooks** [comment]
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-learning&entry.867055334=https://github.com/openedx/openedx-learning/pull/55&entry.1484655318=https://github.com/pomegranited&entry.752974735=2023-06-01+07:16&entry.1917517419=2023-06-13+18:55&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
